### PR TITLE
feat: Add Auto Select Batch Feature in Stock Settings and Delivery Notes, Add coa_batch field in batch doctype

### PR DESCRIPTION
**Pull request by rvpasari**
https://app.asana.com/0/1192407975826276/1200068587152761

### CODE REVIEW

**Comment by sahil28297** [comment]
@rvpasari please resolve the merge conflicts. Also why is the PR raised on bs-mar-sprint-1 branch?

File: erpnext/patches/v13_0/set_coa_batch_in_batch.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import frappe
+
+def execute():
+    """
+    Set coa batch value in batch document if batch and coa batch is set in package tag
+    """
+    frappe.reload_doc("stock", "doctype", "batch")
+    package_tags = frappe.get_all("Package Tag",
+        filters = [{'batch_no': ["is", "set"], "coa_batch_no": ["is", "set"]}],
```
Comment (hrwX) [inline erpnext/patches/v13_0/set_coa_batch_in_batch.py]:
```suggestion
        filters = {'batch_no': ["is", "set"], "coa_batch_no": ["is", "set"]},
```

File: erpnext/patches/v13_0/set_coa_batch_in_batch.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+        fields = ['name', 'batch_no', "coa_batch_no"])
```
Comment (hrwX) [inline erpnext/patches/v13_0/set_coa_batch_in_batch.py]: - we are not using name anywhere right, we can skip that
```suggestion
        fields = ['batch_no', "coa_batch_no"])
```

File: erpnext/stock/doctype/delivery_note/delivery_note.py [hunk 3]
```diff
@@ -162,6 +163,20 @@ def validate_batch_coa(self):
 			if item.coa_batch:
 				item.certificate_of_analysis = frappe.db.get_value("Batch", item.coa_batch, "certificate_of_analysis")
 
+	def update_item_batch_nos(self):
+		"""
+		Add batch nos to items that have coa batch assigned to them.
+		"""
+		stock_settings = frappe.get_doc("Stock Settings")
```
Comment (hrwX) [inline erpnext/stock/doctype/delivery_note/delivery_note.py]: - Use frappe.db.get_value

File: erpnext/stock/doctype/delivery_note/delivery_note.py [hunk 4]
```diff
@@ -162,6 +163,20 @@ def validate_batch_coa(self):
...
+		for item in self.items:
+			#If item has a coa batch and auto assigning batches are enabled in stock settings, we assign a batch to the item in the delivery note
+			if  stock_settings.get("auto_assign_batch_during_delivery", 0) and item.coa_batch and not item.batch_no:
+				#fetch batch with same coa_batch and map quantity
+				valid_batches = frappe.get_all("Batch",
+							filters = {"batch_qty":[">", item.qty], "coa_batch": item.coa_batch}
+				)
```
Comment (hrwX) [inline erpnext/stock/doctype/delivery_note/delivery_note.py:178]: - add `limit=1` instead of fetching all the entries in `get_all`

**Review by hrwX — CHANGES_REQUESTED** [review]

**Pull request closed without merging**